File: haskell/list-ops/src/ListOps.hs
{-# LANGUAGE BangPatterns #-}
module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' f !input ![] = input
foldl' f !input !(x:xs) = foldl' f (f input x) xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr f input [] = input
foldr f input (x:xs) = f x $ foldr f input xs

length :: [a] -> Int
length [] = 0
length xs = sum $ map (\_ -> 1 :: Int) xs

reverse :: [a] -> [a]
reverse [] = []
reverse xs = last xs : reverse (init xs)

map :: (a -> b) -> [a] -> [b]
map _ [] = []
map f (x:xs) = f x : map f xs

filter :: (a -> Bool) -> [a] -> [a]
filter _ [] = []
filter p (x:xs)
    | p x = x : filter p xs
    | otherwise = filter p xs

(++) :: [a] -> [a] -> [a]
[] ++ ys = ys
(x:xs) ++ ys = x : xs ++ ys

concat :: [[a]] -> [a]
concat [] = []
concat l = foldr (++) [] l

File: javascript/list-ops/list-ops.js
class List {
    constructor(list = []) {
        this.values = list;
    }

    length() {
        return this.values.length;
    }

    append(list) {
        list.values.forEach(item => {
            this.values.push(item);
        });
        return this;
    }

    concat(list) {
        this.append(list);
        return this;
    }

    filter(predicate) {
        let results = [];
        while (this.values.length) {
            let item = this.values.shift();
            if (predicate(item)) {
                results.push(item);
            }
        }
        while (results.length) {
            this.values.push(results.shift());
        }
        return this;
    }

    map(map) {
        let results = [];
        while (this.values.length) {
            results.push(map(this.values.shift()));
        }
        while (results.length) {
            this.values.push(results.shift());
        }
        return this;
    }

    foldl(reducer, accumulator) {
        while (this.values.length) {
            accumulator = reducer(this.values.shift(), accumulator);
        }
        return accumulator;
    }

    foldr(reducer, accumulator) {
        while (this.values.length) {
            accumulator = reducer(this.values.pop(), accumulator);
        }
        return accumulator;
    }

    reverse() {
        let results = this.foldr((val, acc) => {
            acc.push(val);
            return acc;
        }, []);
        while (results.length) {
            this.values.push(results.shift());
        }
        return this;
    }
}

module.exports = List;